pkg/protocol/engine/upgrade/orchestrator.go

```diff
@@ -0,0 +1,329 @@
+// of the protocol, rising from the ashes of the previous version.
+//
+// User
+// maybe a more technical and less mythical term
+//
+// ChatGPT
+// Certainly! In that case, you could consider naming the software component the "Upgrade Orchestrator."
+// This name emphasizes its role in orchestrating protocol upgrades, managing the various stages of the process, and
+// ensuring a smooth transition. The term "orchestrator" suggests a conductor-like function, coordinating the upgrade
+// activities and ensuring consensus before activating the changes.
+type Orchestrator struct {
+	evictionMutex syncutils.RWMutex
+	errorHandler  func(error)
+
+	latestSignals           *memstorage.IndexedStorage[iotago.SlotIndex, account.SeatIndex, *prunable.SignaledBlock]
+	upgradeSignalsFunc      func(slot iotago.SlotIndex) *prunable.UpgradeSignals
+	permanentUpgradeSignals *kvstore.TypedStore[iotago.EpochIndex, VersionAndHash]
+
+	setProtocolParametersEpochMappingFunc func(iotago.Version, iotago.EpochIndex) error
+	protocolParametersAndVersionsHashFunc func(iotago.SlotIndex) (iotago.Identifier, error)
+	epochForVersionFunc                   func(iotago.Version) (iotago.EpochIndex, bool)
+
+	apiProvider api.Provider
+	seatManager seatmanager.SeatManager
+
+	module.Module
+}
+
+func NewProvider(opts ...options.Option[Orchestrator]) module.Provider[*engine.Engine, *Orchestrator] {
+	return module.Provide(func(e *engine.Engine) *Orchestrator {
+		o := NewOrchestrator(
+			e.ErrorHandler("upgradegadget"),
+			e.Storage.Permanent.UpgradeSignals(),
+			e.Storage.Prunable.UpgradeSignals,
+			e.Storage.Settings(),
+			e.Storage.Settings().StoreProtocolParametersEpochMapping,
+			e.Storage.Settings().VersionsAndProtocolParametersHash,
+			e.Storage.Settings().EpochForVersion,
+			e.SybilProtection.SeatManager(),
+			opts...,
+		)
+
+		e.Events.BlockGadget.BlockAccepted.Hook(o.trackHighestSupportedVersion)
+		o.TriggerInitialized()
+
+		return o
+	})
+}
+
+func NewOrchestrator(errorHandler func(error),
+	permanentUpgradeSignal kvstore.KVStore,
+	upgradeSignalsFunc func(slot iotago.SlotIndex) *prunable.UpgradeSignals,
+	apiProvider api.Provider,
+	setProtocolParametersEpochMappingFunc func(iotago.Version, iotago.EpochIndex) error,
+	protocolParametersAndVersionsHashFunc func(iotago.SlotIndex) (iotago.Identifier, error),
+	epochForVersionFunc func(iotago.Version) (iotago.EpochIndex, bool),
+	seatManager seatmanager.SeatManager, opts ...options.Option[Orchestrator]) *Orchestrator {
+	return options.Apply(&Orchestrator{
+		errorHandler:            errorHandler,
+		latestSignals:           memstorage.NewIndexedStorage[iotago.SlotIndex, account.SeatIndex, *prunable.SignaledBlock](),
+		permanentUpgradeSignals: kvstore.NewTypedStore[iotago.EpochIndex, VersionAndHash](permanentUpgradeSignal, iotago.EpochIndex.Bytes, iotago.EpochIndexFromBytes, VersionAndHash.Bytes, VersionAndHashFromBytes),
```

Comment:
you can remove the generic types here since it can infer it properly from the assignment